<a href="https://colab.research.google.com/github/ssook1222/Data-Analysis-Crush/blob/main/1%ED%9A%8C%EB%8F%85/1%EC%98%81%EC%97%AD/task1_108.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv")
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

### Question 91

price_range 값이 3인 그룹에서 상관관계가 2번째로 높은 두 컬럼과 그 상관계수를 구하여라

In [ ]:
df91 = df[(df.price_range == 3)].corr().unstack().sort_values(ascending = False)
df91.loc[df91!=1].reset_index().iloc[1]

level_0          pc
level_1          fc
0          0.635166
Name: 1, dtype: object

## 비행탑승 경험 만족도 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv")
df.head()

id  Gender   Customer Type  Age   Type of Travel     Class  \
0   55882    Male  Loyal Customer   42  Business travel  Business   
1   31906  Female  Loyal Customer   31  Business travel  Business   
2   85481    Male  Loyal Customer   51  Personal Travel       Eco   
3  102134    Male  Loyal Customer   44  Business travel  Business   
4   31580    Male  Loyal Customer   33  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0              473                      1                                  1   
1             3271                      4                                  4   
2              214                      5                                  0   
3             3097                      1                                  1   
4             3286                      1                                  1   

   Ease of Online booking  ...  Inflight entertainment  On-board service  \
0                       3  ...                       5                 5   
1                       4  ...                       3                 3   
2                       5  ...                       3                 4   
3                       3  ...                       4                 4   
4                       1  ...                       4                 2   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 5                 5                3                 5   
1                 5                 5                3                 4   
2                 4                 4                4                 4   
3                 4                 4                3                 4   
4                 5                 2                5                 1   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            3                           3                       0.0   
1            3                           1                       5.0   
2            3                           2                       0.0   
3            5                           0                       0.0   
4            4                           0                       0.0   

   satisfaction  
0     satisfied  
1     satisfied  
2     satisfied  
3     satisfied  
4     satisfied  

[5 rows x 24 columns]

### Question 92

Arrival Delay in Minutes 컬럼이 결측치인 데이터들 중 ‘neutral or dissatisfied’ 보다 ‘satisfied’의 수가 더 높은 Class는 어디 인가?

In [ ]:
df92 = df[(df['Arrival Delay in Minutes'].isna() == True)]
df92 = df92[['Class','satisfaction']].groupby(['Class','satisfaction'], as_index = False).size().pivot(index='Class',columns='satisfaction')
df92[df92['size']['neutral or dissatisfied'] < df92['size']['satisfied']]

size          
satisfaction neutral or dissatisfied satisfied
Class                                         
Business                          36        76

## 수질 음용성 여부 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/waters/train.csv")
df.head()

ph    Hardness        Solids  Chloramines     Sulfate  Conductivity  \
0  7.918150  214.186611  23823.492888     7.290878  341.173322    411.424483   
1  5.422446  205.266080  18542.957451     5.491963  306.702227    382.080129   
2  7.341547  187.672402  21273.457066     7.784003         NaN    332.084293   
3  9.056245  197.666301  17403.532167     7.688917  337.460176    414.766631   
4  5.039374  142.860598  40829.353167     7.271543         NaN    386.803057   

   Organic_carbon  Trihalomethanes  Turbidity  Potability  
0       19.585002        25.057375   4.028958           0  
1       10.504023        67.493450   2.911751           1  
2       16.842334        55.019151   4.025644           0  
3       15.349869        63.696746   3.319354           0  
4       16.823773        52.297113   4.957420           0

### Question 93

ph값은 상당히 많은 결측치를 포함한다. 결측치를 제외한 나머지 데이터들 중 사분위값 기준 하위 25%의 값들의 평균값은?

In [ ]:
df[(df.ph <= df[(df.ph.isna() == False)].ph.quantile(0.25))].mean().ph

5.057093462441731

## 의료 비용 예측 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv")
df.head()

age     sex     bmi  children smoker     region      charges
0   64  female  39.330         0     no  northeast  14901.51670
1   47  female  27.830         0    yes  southeast  23065.42070
2   52  female  33.300         2     no  southwest  10806.83900
3   33  female  22.135         1     no  northeast   5354.07465
4   30    male  31.400         1     no  southwest   3659.34600

### Question 94

흡연자와 비흡연자 각각 charges의 상위 10% 그룹의 평균의 차이는?

In [ ]:
non_smoker = df[(df.smoker == 'no')]
smoker = df[(df.smoker == 'yes')]


smoker[(smoker.charges >= smoker.charges.quantile(0.9))].charges.mean() - non_smoker[(non_smoker.charges >= non_smoker.charges.quantile(0.9))].charges.mean()

29297.954548156144

## 킹카운티 주거지 가격예측문제 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice//train.csv")
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  8961990160  20150413T000000  567500.0         3        2.5         2080   
1  9455200205  20140604T000000  525000.0         3        2.0         1540   
2  7853220670  20140918T000000  540000.0         3        2.5         2860   
3  3298201170  20141110T000000  350000.0         3        1.0          940   
4  7972604355  20140521T000000  218000.0         3        1.0         1020   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      4556     2.0           0     0  ...      8        2080              0   
1      7800     1.0           0     0  ...      8        1540              0   
2      8935     2.0           0     0  ...      8        2860              0   
3      7811     1.0           0     0  ...      6         940              0   
4      7874     1.0           0     0  ...      7        1020              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1999             0    98074  47.6036 -122.014           1530   
1      2004             0    98125  47.7041 -122.288           1510   
2      2004             0    98065  47.5336 -121.855           2650   
3      1959             0    98008  47.6195 -122.118           1180   
4      1956             0    98106  47.5175 -122.346           1290   

   sqft_lot15  
0        5606  
1        7800  
2        6167  
3        7490  
4        7320  

[5 rows x 21 columns]

### Question 95

bedrooms 의 빈도가 가장 높은 값을 가지는 데이터들의 price의 상위 10%와 하위 10%값의 차이를 구하여라

In [ ]:
df.bedrooms.value_counts()
df[(df.bedrooms == 3)].price.quantile(0.9) - df[(df.bedrooms == 3)].price.quantile(0.1)

505500.0

## 대학원 입학가능성 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv")
df.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0         246        328          110                  4  4.0  2.5  9.02   
1         100        323          113                  3  4.0  4.0  8.88   
2          79        296           95                  2  3.0  2.0  7.54   
3          53        334          116                  4  4.0  3.0  8.00   
4         444        321          114                  5  4.5  4.5  9.16   

   Research  Chance of Admit  
0         1             0.81  
1         1             0.79  
2         1             0.44  
3         1             0.78  
4         1             0.87

### Question 96

Serial No. 컬럼을 제외하고 ‘Chance of Admit’을 종속변수, 나머지 변수를 독립변수라 할때, 랜덤포레스트를 통해 회귀 예측을 할 떄 변수중요도 값을  출력하라 (시드값에 따라 순서는 달라질수 있음)

In [ ]:
df96 = df.drop('Serial No.', axis = 1)
y = df96['Chance of Admit']
x = df96.drop('Chance of Admit', axis = 1)

In [ ]:
from sklearn.ensemble import RandomForestRegressor

rfr = RandomForestRegressor(n_estimators = 3, max_depth = 3)
model = rfr.fit(x, y)

In [ ]:
pd.DataFrame({'importance': model.feature_importances_}, x.columns)

importance
GRE Score            0.254178
TOEFL Score          0.007985
University Rating    0.000000
SOP                  0.000000
LOR                  0.008178
CGPA                 0.729659
Research             0.000000

## 레드 와인 퀄리티 예측 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            5.0              0.38         0.01             1.6      0.048   
1            5.0              0.42         0.24             2.0      0.060   
2            7.1              0.36         0.30             1.6      0.080   
3            7.6              0.29         0.49             2.7      0.092   
4            7.7              0.51         0.28             2.1      0.087   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 26.0                  60.0  0.99084  3.70       0.75   
1                 19.0                  50.0  0.99170  3.72       0.74   
2                 35.0                  70.0  0.99693  3.44       0.50   
3                 25.0                  60.0  0.99710  3.31       0.61   
4                 23.0                  54.0  0.99800  3.42       0.74   

   alcohol  quality  
0     14.0        6  
1     14.0        8  
2      9.4        5  
3     10.1        6  
4      9.2        5

### Question 97

quality 값이 3인 그룹과 8인 데이터그룹의 각 컬럼별 독립변수의 표준편차 값의 차이를 구할때 그값이 가장 큰 컬럼명을 구하여라

In [ ]:
(df[(df.quality == 8 )].std() - df[(df.quality == 3 )].std()).sort_values().tail(1).index[0]

'total sulfur dioxide'

## 약물 분류 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/drug/train.csv")
df.head()

Age Sex      BP Cholesterol  Na_to_K  Drug
0   20   F  NORMAL      NORMAL    9.281     4
1   24   M    HIGH      NORMAL    9.475     1
2   34   M  NORMAL        HIGH   22.456     0
3   17   M  NORMAL      NORMAL   10.832     4
4   72   M     LOW        HIGH   16.310     0

### Question 98

남성들의 연령대별 (10살씩 구분 0 ~ 9세 10 ~ 19세 …) Na_to_K값의 평균값을 구해서 데이터 프레임으로 표현하여라

In [ ]:
def age(x):
  if 0 <= x <= 9:
    return '0대'

  if 10 <= x <= 19:
    return '10대'

  if 20 <= x <= 29:
    return '20대'

  if 30 <= x <= 39:
    return '30대'

  if 40 <= x <= 49:
    return '40대'

  if 50 <= x <= 59:
    return '50대'

  if 60 <= x <= 69:
    return '60대'

  if 70 <= x <= 79:
    return '70대'

df['age_time'] = df.Age.apply(age)
df[['age_time', 'Na_to_K']].groupby('age_time', as_index = False).mean()

age_time    Na_to_K
0      10대  17.603111
1      20대  17.220038
2      30대  15.656500
3      40대  14.985273
4      50대  15.810231
5      60대  16.226045
6      70대  14.413364

## 사기회사 분류 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/audit/train.csv")
df.head()

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  Risk_B  \
0          2.72          14    4.53      0.6   2.718   87.93      0.6  52.758   
1          3.41          16   12.68      0.6   7.608   41.00      0.6  24.600   
2          2.72          11    3.11      0.6   1.866  113.97      0.6  68.382   
3          3.41           8    1.12      0.4   0.448    0.00      0.2   0.000   
4         55.57           9    1.06      0.4   0.424    0.00      0.2   0.000   

    TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  Inherent_Risk  \
0   92.46      5.0  ...     0.4        0   0.2     0.0    4.0        108.362   
1   53.68      5.5  ...     0.4        0   0.2     0.0    4.4         41.936   
2  117.08      5.5  ...     0.4        0   0.2     0.0    4.4         88.832   
3    1.12      5.0  ...     0.4        0   0.2     0.0    2.2          1.848   
4    1.06      5.0  ...     0.4        0   0.2     0.0    2.2          1.824   

   CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  
0           0.4             0.5     21.6724     1  
1           0.4             0.5      8.3872     1  
2           0.4             0.5     17.7664     1  
3           0.4             0.5      0.3696     0  
4           0.4             0.5      0.3648     0  

[5 rows x 27 columns]

### Question 99

데이터의 Risk 값에 따른 score_a와 score_b의 평균값을 구하여라

In [ ]:
df[['Risk', 'Score_A', 'Score_B']].groupby('Risk', as_index = False).mean()

Risk   Score_A   Score_B
0     0  0.262234  0.225532
1     1  0.490164  0.444262

## 센서데이터 동작유형 분류 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/muscle/train.csv")
df.head()

motion_0  motion_1  motion_2  motion_3  motion_4  motion_5  motion_6  \
0      -6.0       8.0       5.0       6.0     -12.0     -27.0       5.0   
1     -32.0      -3.0      -4.0      -5.0      -1.0      16.0       0.0   
2      35.0      -2.0      -1.0       1.0     -25.0     -16.0      -3.0   
3     -15.0      -6.0      -3.0      -5.0      27.0      37.0      -2.0   
4      60.0       3.0       0.0      -7.0      -6.0     -10.0      -6.0   

   motion_7  motion_8  motion_9  ...  motion_55  motion_56  motion_57  \
0      24.0      -8.0      -5.0  ...       -1.0       -3.0       -6.0   
1      -5.0      42.0       2.0  ...      -17.0      -12.0       -1.0   
2     -20.0      10.0       1.0  ...        6.0      -65.0        6.0   
3      -2.0      19.0       5.0  ...        2.0       11.0        0.0   
4       7.0     -11.0      -4.0  ...      -12.0       17.0       -3.0   

   motion_58  motion_59  motion_60  motion_61  motion_62  motion_63  pose  
0       -9.0       -7.0       36.0       68.0        3.0       -3.0     1  
1       -1.0        0.0      -26.0      -11.0       -4.0       -9.0     1  
2      -10.0      -31.0      -19.0       -3.0      -47.0      -17.0     0  
3       -2.0       -4.0       -7.0       15.0        3.0       14.0     1  
4        1.0        2.0       -2.0       20.0       18.0       10.0     0  

[5 rows x 65 columns]

### [!] Question 100

pose값에 따른 각 motion컬럼의 중간값의 가장 큰 차이를 보이는 motion컬럼은 어디이며 그값은?

In [ ]:
t= df.groupby('pose').median().T
dfs = abs(t[0] - t[1]).sort_values().reset_index()
dfs[dfs[0] ==dfs[0].max()]['index'].values

array(['motion_54', 'motion_62'], dtype=object)

## 현대 차량 가격 분류 문제 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")
df.head()

model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0   Tucson  2016  12795       Manual    36263   Diesel      30  61.7   
1      I10  2012   3995       Manual    22089   Petrol      20  61.4   
2      I30  2019  24000       Manual      555   Petrol     145  34.0   
3   Tucson  2017  12995       Manual    32147   Diesel      30  61.7   
4   Tucson  2018  14991    Semi-Auto    32217   Diesel     150  57.6   

   engineSize  
0         1.7  
1         1.2  
2         2.0  
3         1.7  
4         1.7

### Question 101

정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?

In [ ]:
answer = df.loc[df.model.isin(df.model.value_counts().index[:3])].groupby('model', as_index = False).price.mean()
display(answer)

model         price
0      I10   7646.137891
1      I30  11651.821759
2   Tucson  15805.441373

## 당뇨여부판단 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            3      102             74              0        0  29.5   
1            2      144             58             33      135  31.6   
2            5      136             82              0        0   0.0   
3           13      145             82             19      110  22.2   
4            1      117             60             23      106  33.8   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.121   32        0  
1                     0.422   25        1  
2                     0.640   69        0  
3                     0.245   57        0  
4                     0.466   27        0

### Question 102

Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라

In [ ]:
df102 = df.drop('Outcome', axis = 1)

In [ ]:
pd.DataFrame({'values' : df.groupby('Outcome').mean().values[1] - df.groupby('Outcome').mean().values[0]}, df102.columns)

values
Pregnancies                1.574159
Glucose                   29.943995
BloodPressure              3.202079
SkinThickness              2.952033
Insulin                   33.398645
BMI                        4.541437
DiabetesPedigreeFunction   0.143646
Age                        5.866939

## 넷플릭스 주식 데이터

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")
df.head()

Date      Open      High       Low     Close  Adj Close     Volume
0  2002-05-23  1.156429  1.242857  1.145714  1.196429   1.196429  104790000
1  2002-05-24  1.214286  1.225000  1.197143  1.210000   1.210000   11104800
2  2002-05-28  1.213571  1.232143  1.157143  1.157143   1.157143    6609400
3  2002-05-29  1.164286  1.164286  1.085714  1.103571   1.103571    6757800
4  2002-05-30  1.107857  1.107857  1.071429  1.071429   1.071429   10154200

### Question 103

매년 5월달의 open가격의 평균값을 데이터 프레임으로 표현하라

In [ ]:
import datetime as dt

df['new_date'] = pd.to_datetime(df.Date)
df103 = df[(df.new_date.dt.month == 5)]

df103.groupby(df103.new_date.dt.year).Open.mean()

new_date
2002      1.155833
2003      1.641497
2004      4.261143
2005      1.951905
2006      4.163571
2007      3.159351
2008      4.435034
2009      5.790571
2010     14.417071
2011     34.650272
2012     10.523247
2013     31.936429
2014     51.121292
2015     85.057429
2016     92.705715
2017    158.255455
2018    329.779541
2019    359.664548
2020    433.880499
2021    496.923996
Name: Open, dtype: float64